File: internal/productstate/asset_graph_test.go
package productstate_test

import (
	"errors"
	"testing"

	"cyber-code/internal/productstate"
)

func TestProjectAssetGraphKeepsUnresolvedEdgesAndStatuses(t *testing.T) {
	state := productstate.Initial()
	evidence := mustEvent(t, 1, "evidence.committed", map[string]any{"evidence": map[string]any{"id": "evidence-1", "taskId": "task-1", "kind": "network", "summary": "observed", "data": map[string]any{}}}, nil)
	result, err := productstate.Project(state, evidence)
	if err != nil {
		t.Fatal(err)
	}
	state = result.State
	provenance := map[string]any{"kind": "evidence", "evidenceIds": []string{"evidence-1"}}
	node := map[string]any{"id": "asset:target:lab", "kind": "target", "label": "lab", "status": "active", "attributes": map[string]any{}, "provenance": provenance}
	result, err = productstate.Project(state, mustEvent(t, 2, "asset.node.committed", map[string]any{"node": node}, nil))
	if err != nil {
		t.Fatal(err)
	}
	state = result.State
	edge := map[string]any{"id": "edge:unknown", "kind": "routes_to", "sourceId": "asset:target:lab", "targetId": "asset:route:unknown", "directed": true, "provenance": provenance}
	result, err = productstate.Project(state, mustEvent(t, 3, "asset.edge.committed", map[string]any{"edge": edge}, nil))
	if err != nil {
		t.Fatal(err)
	}
	state = result.State
	if state.AssetEdges["edge:unknown"].TargetID != "asset:route:unknown" {
		t.Fatal("unresolved edge was not retained")
	}
	if _, ok := state.AssetNodes["asset:route:unknown"]; ok {
		t.Fatal("projector fabricated an unresolved node")
	}
	result, err = productstate.Project(state, mustEvent(t, 4, "asset.node.status.changed", map[string]any{"nodeId": "asset:target:lab", "status": "revoked", "reason": "rotated"}, nil))
	if err != nil {
		t.Fatal(err)
	}
	if result.State.AssetNodes["asset:target:lab"].Status != "revoked" {
		t.Fatal("node status was not projected")
	}
}

func TestProjectAssetGraphRequiresEvidenceAndRejectsConflicts(t *testing.T) {
	provenance := map[string]any{"kind": "evidence", "evidenceIds": []string{"missing"}}
	node := map[string]any{"id": "asset:target:lab", "kind": "target", "label": "lab", "status": "active", "attributes": map[string]any{}, "provenance": provenance}
	if _, err := productstate.Project(productstate.Initial(), mustEvent(t, 1, "asset.node.committed", map[string]any{"node": node}, nil)); !errors.Is(err, productstate.ErrAssetEvidenceMissing) {
		t.Fatalf("missing evidence error = %v", err)
	}
}

File: packages/protocol/src/asset-graph.test.ts
import { describe, expect, test } from 'vitest';

import { initialProductState, project, validateEvent, type ProductState, type RawProductEvent } from './index';

const raw = (cursor: number, type: string, payload: Record<string, unknown>): RawProductEvent => ({
  schemaVersion: 1,
  eventId: `event-${cursor}`,
  taskId: 'task-1',
  cursor,
  occurredAt: '2026-08-06T00:00:00Z',
  type,
  source: { runtimeId: 'runtime-local' },
  payload,
});

const apply = (state: ProductState, cursor: number, type: string, payload: Record<string, unknown>) => project(state, validateEvent(raw(cursor, type, payload))).state;

const target = {
  id: 'asset:target:juice-shop.lab',
  kind: 'target',
  label: 'juice-shop.lab',
  status: 'active',
  attributes: { hostname: 'juice-shop.lab' },
  provenance: { kind: 'evidence', evidenceIds: ['evidence-1'] },
};
const service = {
  id: 'asset:service:juice-shop.lab:443',
  kind: 'service',
  label: 'HTTPS 443',
  status: 'active',
  attributes: { port: 443, transport: 'tcp' },
  provenance: { kind: 'evidence', evidenceIds: ['evidence-1'] },
};
const edge = {
  id: 'edge:target-service:juice-shop.lab:443',
  kind: 'exposes',
  sourceId: target.id,
  targetId: service.id,
  directed: true,
  provenance: { kind: 'evidence', evidenceIds: ['evidence-1'] },
};

describe('asset graph protocol', () => {
  test('projects evidence-backed nodes and typed causal edges', () => {
    let state = apply(initialProductState(), 1, 'evidence.committed', { evidence: { id: 'evidence-1', taskId: 'task-1', kind: 'network', summary: '443 open', data: { port: 443 } } });
    state = apply(state, 2, 'asset.node.committed', { node: target });
    state = apply(state, 3, 'asset.node.committed', { node: service });
    state = apply(state, 4, 'asset.edge.committed', { edge });

    expect(state.assetNodes).toEqual({ [target.id]: target, [service.id]: service });
    expect(state.assetEdges).toEqual({ [edge.id]: edge });
    expect(Object.isFrozen(state.assetEdges[edge.id])).toBe(true);
  });

  test('retains an edge with an unresolved endpoint without fabricating a node', () => {
    let state = apply(initialProductState(), 1, 'evidence.committed', { evidence: { id: 'evidence-1', taskId: 'task-1', kind: 'network', summary: 'redirect observed', data: {} } });
    state = apply(state, 2, 'asset.node.committed', { node: target });
    state = apply(state, 3, 'asset.edge.committed', { edge: { ...edge, id: 'edge-unresolved', targetId: 'asset:route:unknown', kind: 'routes_to' } });

    expect(state.assetEdges['edge-unresolved'].targetId).toBe('asset:route:unknown');
    expect(state.assetNodes['asset:route:unknown']).toBeUndefined();
  });

  test('deduplicates identical asset identities and rejects conflicting reuse', () => {
    let state = apply(initialProductState(), 1, 'evidence.committed', { evidence: { id: 'evidence-1', taskId: 'task-1', kind: 'network', summary: '443 open', data: {} } });
    state = apply(state, 2, 'asset.node.committed', { node: target });
    const same = apply(state, 3, 'asset.node.committed', { node: target });
    expect(same.assetNodes).toEqual(state.assetNodes);
    expect(() => apply(same, 4, 'asset.node.committed', { node: { ...target, label: 'conflict' } })).toThrow('asset_node_conflict');

    state = apply(same, 4, 'asset.edge.committed', { edge });
    const sameEdge = apply(state, 5, 'asset.edge.committed', { edge });
    expect(sameEdge.assetEdges).toEqual(state.assetEdges);
    expect(() => apply(sameEdge, 6, 'asset.edge.committed', { edge: { ...edge, kind: 'routes_to' } })).toThrow('asset_edge_conflict');
  });

  test('keeps revoked and deleted nodes inspectable', () => {
    let state = apply(initialProductState(), 1, 'evidence.committed', { evidence: { id: 'evidence-1', taskId: 'task-1', kind: 'credential', summary: 'credential observed', data: {} } });
    state = apply(state, 2, 'asset.node.committed', { node: { ...target, kind: 'credential' } });
    state = apply(state, 3, 'asset.node.status.changed', { nodeId: target.id, status: 'revoked', reason: 'rotated' });
    expect(state.assetNodes[target.id]).toMatchObject({ status: 'revoked', statusReason: 'rotated' });
    state = apply(state, 4, 'asset.node.status.changed', { nodeId: target.id, status: 'deleted', reason: 'removed_from_scope' });
    expect(state.assetNodes[target.id]).toMatchObject({ status: 'deleted', statusReason: 'removed_from_scope' });
  });

  test('requires auditable provenance and valid graph envelopes', () => {
    for (const event of [
      raw(1, 'asset.node.committed', { node: { ...target, provenance: { kind: 'evidence', evidenceIds: [] } } }),
      raw(1, 'asset.edge.committed', { edge: { ...edge, provenance: { kind: 'human', annotationId: '', author: 'operator' } } }),
      raw(1, 'asset.edge.committed', { edge: { ...edge, sourceId: edge.targetId } }),
      raw(1, 'asset.node.status.changed', { nodeId: target.id, status: 'active', reason: 'rollback' }),
    ]) expect(() => validateEvent(event)).toThrow('invalid_event');
  });

  test('rejects evidence provenance that does not reference committed Evidence', () => {
    expect(() => apply(initialProductState(), 1, 'asset.node.committed', { node: target })).toThrow('asset_evidence_missing');
  });
});
